napari/viewer.py

```diff
@@ -26,6 +29,10 @@ class Viewer(ViewerModel):
         Dimension names. by default they are labeled with sequential numbers
     show : bool, optional
         Whether to show the viewer after instantiation. by default True.
+    icon_bugfix: bool
+        If use bugfix to show proper icon when running as script on Windows.
+        May want to disable when embed napari as viewer in own application
```

Comment:
@Czaki, I know you're thinking ahead to embedding napari in your own application, but `napari.Viewer` is pretty much our _main_ offering, and I think we want to be very protective about adding parameters.  I'll defer to @jni and @sofroniewn, but I don't think this parameter is going to make the cut.

Reply:
But forcing this cptypes overwrite all time it may break behaviour of napari based (using docking system) applications on windows. 

In my case I use `ViewerModel` so there is no problem for me. 

You sell napari as viewer, not complex stand-alone application.

Reply:
I'm not saying you have to force it, but I am politely requesting that you consider alternatives that don't require adding a parameter to our main product.  perhaps a module level variable?  etc...

I'm really hoping this is a lost-in-translation thing... but I have to tell you that it comes off as extremely rude for you to tell us how we "sell our application" as justification for rejecting a request.  We _greatly_ value your input, but please also respect that we have lots of different users and use cases, and we like to consider all of them before simply accepting a proposed change.  There are a _lot_ of people who will never embed napari, but may look at our Viewer parameters and say "What is this `icon_bugfix` thing?  do I need to use it?... What's the bug?"  In general, please try to be open to these discussions.

Reply:
(I'm sorry if I overreacted or was too harsh there, I'd just really like us to consider all alternatives and would like you to help us in that process)

Reply:
@Czaki thanks for your input, both here and in the other PRs where you have contributed.

I think @tlambert03 overreacted a bit because the scope of napari might not be immediately clear from the README or existing uses. napari is many things and one of them is a Python viewer, indeed — and from that angle, we actually want to be *more* protective of the API signature, not less.

In the future, we do hope to release a napari standalone application on which other applications can be built as plugins. So napari will *also* be an application. You can see our public short-term roadmap [here](https://napari.org/docs/developers/ROADMAP_0_3.html).

At any rate, this is not to dismiss your concern — just to say that this is not the solution we would like to address that concern. Something like an environment variable, eg NAPARI_DISABLE_WIN_ICON, that we can check, would be much less obtrusive but still accessible for application developers. 

Again, thanks for working with us to come up with a solution that works for everyone! What do you think of the environment variable approach?

Reply:
I do not think that enviroment variable is good solution (it is global and rather for user than for programmer)

> I'm really hoping this is a lost-in-translation thing... but I have to tell you that it comes off as extremely rude for you to tell us how we "sell our application" as justification for rejecting a request.

I base on current description of napari. Maybe it is outdated 

> Napari viewer is a fast, interactive, multi-dimensional image viewer for Python. It’s designed for browsing, annotating, and analyzing large multi-dimensional images. It’s built on top of Qt (for the GUI), vispy (for performant GPU-based rendering), and the scientific Python stack (numpy, scipy).

But for me some part of code which will group different napari based solution in one group on windows which need code edition to disable is not a good idea. In first try I put it in qt loop initialisation to add programmer possibility to redefine _identifier_. But this solution does not work with napari launched from `ipython --gui qt`. 

I do not see other option than put it in Viewer constructor because it is always executed before create napari window, but for me this need to be easy to disable. Way which i see are:

 * Constructor argument - easiest to found, but may also some basic user may be not sure what it is. 
 * Environment variable - strongly against. As mentioned before it is rather for end user  than for programmer
 * Module level variable - in which module?
 * Class level variable  - not visible at firs moment but easy to control (overwrite before object creation, sublclass with new value) 
 * Check if current object instance of subclass and then not apply bugfix?  

One thing which can be done is to check if control variable is string and if is then use it value instead of predefined one.

Reply:
@tlambert03 @jni

Reply:
@Czaki 

> Environment variable - strongly against. As mentioned before it is rather for end user than for programmer

Well, note that a programmer can execute:

```python
import os

os.environ['NAPARI_DISABLE_WIN_ICON'] = '1'
```

before importing napari. So I don't see environment variables as a big barrier to generalisation here.

> Module level variable - in which module?

Could do. `napari._qt.WIN_ICON = True` by default, but app developers can set it to False if needed. That works for me. @sofroniewn?

The key thing for me is that this is a private, hidden from view thing, not something for our public API. It's an ugly hack and we don't expose ugly hacks — but maybe we allow ugly hacks to patch our ugly hacks. =P

Reply:
> But forcing this cptypes overwrite all time it may break behaviour of napari based (using docking system) applications on windows.
>
> In my case I use ViewerModel so there is no problem for me.

Sounds like disabling this functionality is not currently needed and this is more about potential bugs rather than current usage so I think we should remove the kwarg and just have that first if statement always run for now.

Can you make that change @Czaki?

We can always add in the ability to turn off the functionality in the future if many people are requesting it or it is clearly blocking a major usage mode. 

Stepping back I think that apps that are looking to consume the napari viewer, embed it directly in their own Qt application, and personalize the icon, should be importing the `QtViewer` directly - see #1122 for proposals to make that easier, rather than importing the `Viewer` directly, and so I don't see this functionality needed for that use case.

Hope this is an ok resolution.

Reply:
> Well, note that a programmer can execute:
> 
> ```python
> import os
> 
> os.environ['NAPARI_DISABLE_WIN_ICON'] = '1'
> ```
> 
> before importing napari. So I don't see environment variables as a big barrier to generalisation here.

This is not pythonic way. In exception to really special cases there should be no code execution before import ends. 


> Stepping back I think that apps that are looking to consume the napari viewer, embed it directly in their own Qt application, and personalize the icon, should be importing the `QtViewer` directly - see #1122 for proposals to make that easier, rather than importing the `Viewer` directly, and so I don't see this functionality needed for that use case.

I decide to embed viewer but You suggest to use docking instead of `QtViewer` and `ViewerModel` in same isue which you pointed  https://github.com/napari/napari/pull/1122#issuecomment-611595540.

Reply:
Maybe this? removed kwargs, but allow control by user

Reply:
Can you also drop the `icon_bugfix` doc string. I'll let @jni then review